src/opm/parser/eclipse/EclipseState/Runspec.cpp

```diff
@@ -16,19 +16,110 @@
+#include <opm/parser/eclipse/Parser/ParserKeywords/O.hpp>
+#include <opm/parser/eclipse/Parser/ParserKeywords/P.hpp>
 #include <opm/parser/eclipse/Parser/ParserKeywords/S.hpp>
 #include <opm/parser/eclipse/Parser/ParserKeywords/T.hpp>
 #include <opm/parser/eclipse/Parser/ParserKeywords/W.hpp>
-#include <opm/parser/eclipse/EclipseState/Runspec.hpp>
+
 #include <opm/common/OpmLog/OpmLog.hpp>
 
+#include <ostream>
+#include <stdexcept>
+#include <type_traits>
+
+namespace {
+    Opm::Phases inferActivePhases(const Opm::Deck& deck)
+    {
+        return {
+            deck.hasKeyword<Opm::ParserKeywords::OIL>(),
+            deck.hasKeyword<Opm::ParserKeywords::GAS>(),
+            deck.hasKeyword<Opm::ParserKeywords::WATER>(),
+            deck.hasKeyword<Opm::ParserKeywords::SOLVENT>(),
+            deck.hasKeyword<Opm::ParserKeywords::POLYMER>(),
+            deck.hasKeyword<Opm::ParserKeywords::THERMAL>(),
+            deck.hasKeyword<Opm::ParserKeywords::POLYMW>(),
+            deck.hasKeyword<Opm::ParserKeywords::FOAM>(),
+            deck.hasKeyword<Opm::ParserKeywords::BRINE>(),
+            deck.hasKeyword<Opm::ParserKeywords::PVTSOL>()
+        };
+    }
+
+    Opm::SatFuncControls::ThreePhaseOilKrModel
+    inferThreePhaseOilKrModel(const Opm::Deck& deck)
+    {
+        using KroModel = Opm::SatFuncControls::ThreePhaseOilKrModel;
+
+        if (deck.hasKeyword<Opm::ParserKeywords::STONE1>())
+            return KroModel::Stone1;
+
+        if (deck.hasKeyword<Opm::ParserKeywords::STONE>() ||
+            deck.hasKeyword<Opm::ParserKeywords::STONE2>())
+            return KroModel::Stone2;
+
+        return KroModel::Default;
+    }
+
+    Opm::SatFuncControls::KeywordFamily
+    inferKeywordFamily(const Opm::Deck& deck)
+    {
+        const auto phases = inferActivePhases(deck);
+        const auto wat    = phases.active(Opm::Phase::WATER);
+        const auto oil    = phases.active(Opm::Phase::OIL);
+        const auto gas    = phases.active(Opm::Phase::GAS);
+
+        const auto threeP = gas && oil && wat;
+        const auto twoP = (!gas && oil && wat) || (gas && oil && !wat);
+
+        const auto family1 =       // SGOF/SLGOF and/or SWOF
+            (gas && (deck.hasKeyword<Opm::ParserKeywords::SGOF>() ||
+                     deck.hasKeyword<Opm::ParserKeywords::SLGOF>())) ||
+            (wat && deck.hasKeyword<Opm::ParserKeywords::SWOF>());
+
+        // note: we allow for SOF2 to be part of family1 for threeP +
+        // solvent simulations.
+
+        const auto family2 =      // SGFN, SOF{2,3}, SWFN
+            (gas && deck.hasKeyword<Opm::ParserKeywords::SGFN>()) ||
+            (oil && ((threeP && deck.hasKeyword<Opm::ParserKeywords::SOF3>()) ||
+                     (twoP && deck.hasKeyword<Opm::ParserKeywords::SOF2>()))) ||
+            (wat && deck.hasKeyword<Opm::ParserKeywords::SWFN>());
+
+        if (gas &&
+            deck.hasKeyword<Opm::ParserKeywords::SGOF>() &&
```

Comment:
In situations like this - throwing a fatal exception based on invalid deck input I try to output the keywordlocation; i.e. something like this:
```C+++
{
     const auto& sgof_location = deck.getKeyword<Opm::ParserKeywords::SGOF>().location();
     const auto& slgof_location = deck.getKeyword<Opm::ParserKeywords::SLGOF>().location();

     msg = fmt::format("Tables SGOF and SLGOF are mutually exclusive\n"
                        "SGOF: line {} in {}\n"
                        "SLGOF: line {} in {}", sgof_location.lineno, sgof_location.filename, 
                                                slgof_location.lineno, slgof_location.filename);
    throw std::invalid_argument(msg);
``` 
}

Reply:
> throwing a fatal exception based on invalid deck input I try to output the keyword location;

We may even add `prohibits` clauses to the keywords to ensure we never get here.  Do those emit location information? If not I'll do what you suggest.

Reply:
> We may even add prohibits clauses to the keywords to ensure we never get here.

yes - that is probably the best solution.

> Do those emit location information? 

Yes [they do](https://github.com/OPM/opm-common/blob/master/src/opm/parser/eclipse/Parser/Parser.cpp#L590)